text/0000-command-ergonomics.md

```diff
@@ -433,6 +434,29 @@ for process invocation and output handling.
 Perhaps printing the command arguments is overly verbose,
 and we should print only the command name.
 
+## `cause` in `Error` impl of `ProcessError`
+
+Because maybe several things went wrong, ever providing a `Some` `cause`
+would involve prioritising multiple possible problems.
```

Comment:
Can you give an example where multiple different things could simultaneously go wrong? It *seems* like `spawn_error`, `communication_error`, and `utf8_error` should be mutually exclusive.

Reply:
I think the following can all occur in the same execution:
 * The application calls `.stderr(piped)` and `.read_stdout()`
 * Spawn succeeds
 * The subprocess produces some invalid-UTF output
 * Then, something in the kernel goes very badly awry (or, some other thing in our Rust program wrecks our fd) and `read()` on one of the pipes fails
 * The subprocess prints some stderr (which the application has said to treat as an error)
 * The subprocess then crashes.

We must report *at least* all of: the error from `read`; the stderr output (which might well be very helpful for diagnosis); and the process's exit status.

I think we should eagerly do a unicode validity check whenever we captured the stdout, and record the unicode conversion error in the `ProcessError`.  That way a caller who calls your `.permit_status()` and gets `Ok` can know that the stdout output is UTF-8; conversely, if the command unexpectedly printed non-UTF-8, the caller can conveniently report the command line etc.

Or to put it another way, we shouldn't "short circuit" some checks and report only a subset of the errors.